Index: 3rdparty/libjpeg/CMakeLists.txt
===================================================================
--- 3rdparty/libjpeg/CMakeLists.txt	(revision 4591)
+++ 3rdparty/libjpeg/CMakeLists.txt	(working copy)
@@ -42,6 +42,11 @@
     set_source_files_properties(jcdctmgr.c PROPERTIES COMPILE_FLAGS "-O1")
 endif()
 
+IF(WIN32 AND CV_ICC)
+set_source_files_properties(jcdctmgr.c PROPERTIES COMPILE_FLAGS "-Od")
+set_source_files_properties(jddctmgr.c PROPERTIES COMPILE_FLAGS "-Od")
+ENDIF()
+
 set_target_properties(${the_target}
 	PROPERTIES OUTPUT_NAME "${the_target}"
 	DEBUG_POSTFIX "${OPENCV_DEBUG_POSTFIX}"
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 4591)
+++ CMakeLists.txt	(working copy)
@@ -640,21 +640,12 @@
                    link_directories("${TBB_LIB_DIR}")
                    set(OPENCV_LINKER_LIBS ${OPENCV_LINKER_LIBS} tbb)
                else()
-                   if(${CMAKE_SYSTEM_PROCESSOR} MATCHES amd64*)
+                   if(CMAKE_CL_64)
                        set(X86_64 1)
                    endif()
-                   if(${CMAKE_SYSTEM_PROCESSOR} MATCHES x86_64*)
-                       set(X86_64 1)
-                   endif()
-                   if(${CMAKE_SYSTEM_PROCESSOR} MATCHES i686*)
+                   if(NOT CMAKE_CL_64)
                        set(X86 1)
                    endif()
-                   if(${CMAKE_SYSTEM_PROCESSOR} MATCHES i386*)
-                       set(X86 1)
-                   endif()
-                   if(${CMAKE_SYSTEM_PROCESSOR} MATCHES x86*)
-                       set(X86 1)
-                   endif()
 
                    set(_TBB_LIB_PATH "${TBB_INCLUDE_DIR}/../lib")
                    if (X86_64)
Index: modules/legacy/src/3dtracker.cpp
===================================================================
--- modules/legacy/src/3dtracker.cpp	(revision 4591)
+++ modules/legacy/src/3dtracker.cpp	(working copy)
@@ -331,7 +331,7 @@
     int i;
     int x, y;
     CvPoint prev_pt = { 0, 0 };
-    static const CvScalar rgb_colors[] = {
+    static CvScalar rgb_colors[] = {
         {{0,0,255}},
         {{0,128,255}},
         {{0,200,200}},
@@ -339,10 +339,10 @@
         {{200,200,0}},
         {{255,0,0}},
         {{255,0,255}} };
-    static const CvScalar gray_colors[] = {
+    static CvScalar gray_colors[] = {
         {{80}}, {{120}}, {{160}}, {{200}}, {{100}}, {{140}}, {{180}}
     };
-    const CvScalar* colors = img->nChannels == 3 ? rgb_colors : gray_colors;
+    CvScalar* colors = img->nChannels == 3 ? rgb_colors : gray_colors;
 
     CvScalar color = colors[0];
     for (y = 0, i = 0; y < etalon_size.height; y++)
Index: OpenCVModule.cmake
===================================================================
--- OpenCVModule.cmake	(revision 4591)
+++ OpenCVModule.cmake	(working copy)
@@ -6,6 +6,7 @@
 
     include_directories("${CMAKE_CURRENT_SOURCE_DIR}/include"
                         "${CMAKE_CURRENT_SOURCE_DIR}/src"
+						"${CMAKE_CURRENT_SOURCE_DIR}/modules/features2d/include"
                         "${CMAKE_CURRENT_BINARY_DIR}")
     
     foreach(d ${ARGN})
